File: Refresh.py


import sqlite3
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
import threading
import logging
import sys
import argparse
from multiprocessing import cpu_count
from tqdm import tqdm
import time
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

DB_PATH = "data/DB/highscore.db"

# Get all player IDs and usernames from airport_highscore
PLAYER_IDS_QUERY = "SELECT userId, userName FROM airport_highscore"
REFRESH_STATS_QUERY = "UPDATE airport_highscore SET userName = ?, userXP = ?, aircraftCount = ?, destinations = ?, battleWins = ? WHERE userId = ?"

API_PROFILE_TEMPLATE = "https://api.skycards.oldapes.com/users/pub/{}"
token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIwMTljMmVmMi0xMDEyLTcwNjEtYWUzNS0wNjlmMThhZTQyYjIiLCJqdGkiOiJIaVdiZlVFM0lnMjBONkhqYnVWT2pfeGFZa2VlNDRaSVZHcFpXX2htSGVNIiwiaWF0IjoxNzcwODI1MzM3LCJleHAiOjQ5MjY1ODUzMzd9.-fOLPEtuvKqOuLnk2EZT8f_Lf-ymMIp4_vjnVgqzNEo"
HEADERS = {
    "Authorization": f"Bearer {token}",
    "accept": "application/json",
    "accept-encoding": "gzip",
    "content-type": "application/json",
    "Host": "api.skycards.oldapes.com",
    "User-Agent": "okhttp/4.12.0",
    "x-client-version": "2.0.24",
}

logging.basicConfig(level=logging.INFO, format='%(asctime)s [%(levelname)s] %(message)s')

# Thread-local storage for database connections and HTTP sessions
_thread_local = threading.local()

# Global batch update queue
update_queue = []
queue_lock = threading.Lock()
BATCH_SIZE = 500  # Commit every 500 updates

def get_db_connection():
    """Get or create a database connection for the current thread."""
    if not hasattr(_thread_local, 'db_conn'):
        conn = sqlite3.connect(DB_PATH, timeout=30, check_same_thread=False)
        conn.row_factory = sqlite3.Row
        try:
            conn.execute("PRAGMA journal_mode=WAL;")
        except Exception:
            pass
        try:
            conn.execute("PRAGMA busy_timeout = 30000;")
        except Exception:
            pass
        _thread_local.db_conn = conn
    return _thread_local.db_conn

def get_http_session():
    """Get or create an HTTP session for the current thread with connection pooling and retry strategy."""
    if not hasattr(_thread_local, 'http_session'):
        session = requests.Session()
        
        # Configure retry strategy with exponential backoff
        retry_strategy = Retry(
            total=3,  # total number of retries
            backoff_factor=0.5,  # exponential backoff: 0.5s, 1s, 2s
            status_forcelist=[429, 500, 502, 503, 504],  # retry on these status codes
            allowed_methods=["GET"]
        )
        
        # Create HTTPAdapter with connection pooling and retry strategy
        adapter = HTTPAdapter(
            max_retries=retry_strategy,
            pool_connections=100,
            pool_maxsize=100
        )
        session.mount("https://", adapter)
        session.mount("http://", adapter)
        
        _thread_local.http_session = session
    return _thread_local.http_session

def close_db_connection():
    """Close database connection and HTTP session for current thread."""
    if hasattr(_thread_local, 'db_conn'):
        _thread_local.db_conn.close()
        del _thread_local.db_conn
    if hasattr(_thread_local, 'http_session'):
        _thread_local.http_session.close()
        del _thread_local.http_session

def batch_update_stat(user_data, user_id):
    """Queue user data for batch update."""
    if user_data is None:
        return
    global update_queue
    with queue_lock:
        # user_data is a tuple: (userName, userXP, aircraftCount, destinations, battleWins)
        update_queue.append(user_data + (user_id,))
        if len(update_queue) >= BATCH_SIZE:
            flush_batch_updates()

def flush_batch_updates():
    """Flush all queued updates to database."""
    global update_queue
    if not update_queue:
        return
    
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.executemany(REFRESH_STATS_QUERY, update_queue)
        conn.commit()
        count = len(update_queue)
        update_queue = []
        return count
    except Exception as e:
        logging.error(f"Failed to batch update XP: {e}")
        update_queue = []
        return 0

def refresh_stats(person_id, username):
    """Fetch user stats from API with automatic retry on failure."""
    url = API_PROFILE_TEMPLATE.format(person_id)
    try:
        session = get_http_session()
        resp = session.get(url, headers=HEADERS, timeout=5)  # Reduced timeout to 5s
        resp.raise_for_status()
        data = resp.json()
        # Extract all user stats from API response
        # Try to get username from API, but use existing username only if API call fails completely
        user_name = data.get("name") or data.get("userName") or data.get("displayName") or ""
        user_xp = data.get("userXP") or data.get("xp") or 0
        aircraft_ct = (
            data.get("aircraftCount")
            or data.get("aircraft_count")
            or data.get("aircrafts")
            or data.get("aircraftsCount")
            or data.get("numAircraftModels")
            or 0
        )
        dest_ct = (
            data.get("destinations")
            or data.get("destinationsCount")
            or data.get("destinationCount")
            or data.get("numDestinations")
            or 0
        )
        battle_wins = data.get("battleWins") or data.get("battlesWon") or data.get("numBattleWins") or 0
        return (user_name, user_xp, aircraft_ct, dest_ct, battle_wins)
    except Exception as e:
        # If API call fails after retries, keep existing username
        logging.debug(f"Failed to fetch data for user {person_id}: {e}")
        return (username, 0, 0, 0, 0)


def main(num_threads):
    conn = sqlite3.connect(DB_PATH, timeout=30, check_same_thread=False)
    try:
        conn.execute("PRAGMA journal_mode=WAL;")
    except Exception:
        pass
    try:
        conn.execute("PRAGMA busy_timeout = 30000;")
    except Exception:
        pass
    
    cursor = conn.cursor()
    cursor.execute(PLAYER_IDS_QUERY)
    player_ids = cursor.fetchall()
    total = len(player_ids)
    
    conn.close()
    
    logging.info(f"Starting refresh of {total} players with {num_threads} threads")

    def worker(row):
        (userId, userName) = row
        user_data = refresh_stats(userId, userName)
        # Queue update for batch processing
        batch_update_stat(user_data, userId)

    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        with tqdm(total=total, desc="Refreshing players", unit="player", ncols=80) as pbar:
            futures = {executor.submit(worker, row): row for row in player_ids}
            for fut in as_completed(futures):
                fut.result()  # propagate exceptions
                pbar.update(1)
    
    # Flush any remaining updates
    with queue_lock:
        remaining = flush_batch_updates()
    
    # Clean up thread-local connections
    close_db_connection()
    
    logging.info(f"Refresh complete! Processed {total} players")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Refresh user stats with multithreading")
    # Use CPU count * 2 as default (good balance for I/O-bound operations)
    default_threads = max(16, cpu_count() * 2)
    parser.add_argument('--threads', type=int, default=default_threads, help=f'Number of threads to use (default: {default_threads})')
    args = parser.parse_args()
    main(args.threads)

File: src/refresh.rs
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::Semaphore;
use anyhow::Result;
use clap::Parser;
use indicatif::ProgressBar;
use reqwest::Client;
use rusqlite::params;

const API_USER_TEMPLATE: &str = "https://api.skycards.oldapes.com/users/pub/";

const TOKEN: &str = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIwMTljMmVmMi0xMDEyLTcwNjEtYWUzNS0wNjlmMThhZTQyYjIiLCJqdGkiOiJIaVdiZlVFM0lnMjBONkhqYnVWT2pfeGFZa2VlNDRaSVZHcFpXX2htSGVNIiwiaWF0IjoxNzcwODI1MzM3LCJleHAiOjQ5MjY1ODUzMzd9.-fOLPEtuvKqOuLnk2EZT8f_Lf-ymMIp4_vjnVgqzNEo";

const BATCH_SIZE: usize = 500;
const DB_PATH: &str = "data/DB/highscore.db";

#[derive(Parser)]
#[command(about = "Refresh player stats from the Skycards API")]
struct Args {
    #[arg(long, default_value = "12", help = "Number of concurrent workers")]
    workers: usize,
}

#[derive(serde::Deserialize, Debug)]
struct UserProfile {
    #[serde(alias = "userXP", alias = "xp")]
    user_xp: Option<i32>,
    #[serde(alias = "aircraftCount", alias = "aircraft_count", alias = "aircrafts", alias = "aircraftsCount", alias = "numAircraftModels")]
    aircraft_count: Option<i32>,
    #[serde(alias = "destinations", alias = "destinationsCount", alias = "destinationCount", alias = "numDestinations")]
    destinations: Option<i32>,
    #[serde(alias = "battleWins", alias = "battlesWon", alias = "numBattleWins")]
    battle_wins: Option<i32>,
    #[serde(alias = "name", alias = "userName", alias = "displayName")]
    user_name: Option<String>,
}

#[derive(Clone)]
struct PlayerUpdate {
    user_id: String,
    user_name: String,
    xp: i32,
    aircraft_count: i32,
    destinations: i32,
    battle_wins: i32,
}

fn create_client() -> Client {
    let mut headers = reqwest::header::HeaderMap::new();
    headers.insert(
        reqwest::header::AUTHORIZATION,
        format!("Bearer {}", TOKEN).parse().unwrap(),
    );
    headers.insert(
        reqwest::header::ACCEPT,
        "application/json".parse().unwrap(),
    );
    headers.insert(
        reqwest::header::CONTENT_TYPE,
        "application/json".parse().unwrap(),
    );
    headers.insert(
        "Host",
        "api.skycards.oldapes.com".parse().unwrap(),
    );
    headers.insert(
        "x-client-version",
        "2.0.24".parse().unwrap(),
    );

    reqwest::Client::builder()
        .default_headers(headers)
        .timeout(Duration::from_secs(5))
        .build()
        .unwrap()
}

fn get_player_ids() -> Result<Vec<(String, String)>> {
    let conn = rusqlite::Connection::open(DB_PATH)?;
    let mut stmt = conn.prepare("SELECT userId, userName FROM airport_highscore")?;

    let players = stmt
        .query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
        })?
        .collect::<Result<Vec<_>, _>>()?;

    Ok(players)
}

async fn fetch_user_stats(client: &Client, user_id: &str, user_name: &str) -> PlayerUpdate {
    let url = format!("{}{}", API_USER_TEMPLATE, user_id);
    
    match client.get(&url).send().await {
        Ok(resp) => match resp.json::<UserProfile>().await {
            Ok(profile) => PlayerUpdate {
                user_id: user_id.to_string(),
                user_name: profile.user_name.unwrap_or_else(|| user_name.to_string()),
                xp: profile.user_xp.unwrap_or(0),
                aircraft_count: profile.aircraft_count.unwrap_or(0),
                destinations: profile.destinations.unwrap_or(0),
                battle_wins: profile.battle_wins.unwrap_or(0),
            },
            Err(_) => PlayerUpdate {
                user_id: user_id.to_string(),
                user_name: user_name.to_string(),
                xp: 0,
                aircraft_count: 0,
                destinations: 0,
                battle_wins: 0,
            },
        },
        Err(_) => PlayerUpdate {
            user_id: user_id.to_string(),
            user_name: user_name.to_string(),
            xp: 0,
            aircraft_count: 0,
            destinations: 0,
            battle_wins: 0,
        },
    }
}

fn batch_update_players(updates: Vec<PlayerUpdate>) -> Result<()> {
    let mut conn = rusqlite::Connection::open(DB_PATH)?;
    let tx = conn.transaction()?;

    for update in updates {
        tx.execute(
            "UPDATE airport_highscore SET userName = ?1, userXP = ?2, aircraftCount = ?3, destinations = ?4, battleWins = ?5 WHERE userId = ?6",
            params![
                &update.user_name,
                update.xp,
                update.aircraft_count,
                update.destinations,
                update.battle_wins,
                &update.user_id
            ],
        )?;
    }

    tx.commit()?;
    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    env_logger::init();
    let args = Args::parse();

    let players = get_player_ids()?;
    let total = players.len();
    println!("Starting refresh of {} players with {} workers\n", total, args.workers);

    let pbar = ProgressBar::new(total as u64);
    pbar.set_style(
        indicatif::ProgressStyle::default_bar()
            .template("  {spinner:.green} [{bar:35.cyan/blue}] {pos:>4}/{len:4} players")
            .unwrap()
            .progress_chars("▓▒░")
            .tick_strings(&["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"]),
    );

    let client = create_client();
    let semaphore = Arc::new(Semaphore::new(args.workers));
    let mut tasks = Vec::new();

    let (tx, mut rx) = tokio::sync::mpsc::channel::<PlayerUpdate>(BATCH_SIZE);
    let mut batch = Vec::new();

    for (user_id, user_name) in players {
        let client = client.clone();
        let semaphore = Arc::clone(&semaphore);
        let tx = tx.clone();
        let pbar = pbar.clone();

        let task = tokio::spawn(async move {
            let _permit = semaphore.acquire().await.unwrap();
            let update = fetch_user_stats(&client, &user_id, &user_name).await;
            let _ = tx.send(update).await;
            pbar.inc(1);
        });

        tasks.push(task);
    }

    drop(tx);

    // Spawn batch processor
    let batch_task = tokio::spawn(async move {
        while let Some(update) = rx.recv().await {
            batch.push(update);
            if batch.len() >= BATCH_SIZE {
                if let Err(e) = batch_update_players(batch.drain(..).collect()) {
                    eprintln!("Failed to batch update: {}", e);
                }
            }
        }

        // Final batch
        if !batch.is_empty() {
            if let Err(e) = batch_update_players(batch) {
                eprintln!("Failed to batch update (final): {}", e);
            }
        }
    });

    // Wait for all tasks to complete
    for task in tasks {
        let _ = task.await;
    }

    batch_task.await?;
    pbar.finish_with_message("✓ Complete");

    println!("\n✓ Refresh complete! Processed {} players", total);

    Ok(())
}
